src/boot.lisp

```diff
@@ -409,66 +411,126 @@
     `(multiple-value-call (lambda ,gvars ,@setqs)
        ,@form)))
 
+;;; types utils
+(defun object-type-code (object) (oget object "dt_Name"))
+(defun set-object-type-code (object tag) (oset tag object "dt_Name"))
+
+;;; for to accurately definition  LIST and CONS forms
+;;; used:
+;;; (typep form
+;;;    (cons
+;;;      (cond ((true-cons-p form) .... code sensetive for cons)
+;;;            (t  ...))))
+;;;
+(defun true-cons-p (form)
+  (%js-try
+   (progn
+     (list-length form)
+     nil)
+   (catch (err)
+     t)))
+
+;;; pure list predicate: (true-list-p (cons 1 2)) => nil
+;;;                      (listp (cons 1 2)) => t
+(defun true-list-p (obj) (and (consp obj) (not (true-cons-p obj))))
 
+;;; types predicate's
 ;;; mop predicate
 (defun mop-object-p (obj)
     (and (consp obj)
-         (eq (oget obj "tagName") :mop-object)
+         (eql (object-type-code obj) :mop-object)
          (= (length obj) 5)))
 
-;;; js-object predicate
+;;; future upgrade
+(defun clos-object-p (object) (eql (object-type-code object) :clos_object))
+
 (defun js-object-p (obj)
   (if (or (sequencep obj)
           (numberp obj)
           (symbolp obj)
           (functionp obj)
+          (characterp obj)
           (packagep obj))
       nil
       t))
 
-;; Incorrect typecase, but used in NCONC.
-(defmacro typecase (x &rest clausules)
+;;; symbol-function from compiler macro
+(defun functionp (f) (functionp f))
+(defun js-null-p (obj) (js-null-p obj))
+(defun js-undefined-p (obj) (js-undefined-p obj))
+
+(defun true () t)
```

Comment:
If this intended to be used for `FFI`?  If so, it should probably go into the `ffi.lisp` file.

Reply:
Done https://github.com/jscl-project/jscl/pull/390/commits/f50ce4f0fd5c403c4ee3f8f82232a86e229efbaa